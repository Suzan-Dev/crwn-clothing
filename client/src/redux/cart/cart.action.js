import CartActionType from './cart.type';

export const ToggleHiddenCart = () => ({
  type: CartActionType.TOGGLE_HIDDEN_CART,
});

export const addItem = (item) => ({
  type: CartActionType.ADD_ITEM,
  payload: item,
});

export const removeItem = (item) => ({
  type: CartActionType.REMOVE_ITEM,
  payload: item,
});

export const removeItemFromCart = (item) => ({
  type: CartActionType.REMOVE_ITEM_FROM_CART,
  payload: item,
});

export const clearCart = () => ({
  type: CartActionType.CLEAR_CART,
});
