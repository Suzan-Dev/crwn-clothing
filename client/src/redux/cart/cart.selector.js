import { createSelector } from 'reselect';

const selectItem = (state) => state.cart;

export const selectCartItem = createSelector(
  [selectItem],
  (cart) => cart.cartItem
);

export const selectHidden = createSelector([selectItem], (cart) => cart.hidden);

export const selectCartItemCount = createSelector(
  [selectCartItem],
  (cartItem) => cartItem.reduce((acc, cartItem) => acc + cartItem.quantity, 0)
);

export const selectCartTotal = createSelector([selectCartItem], (cartItem) =>
  cartItem.reduce(
    (acc, cartItem) => acc + cartItem.quantity * cartItem.price,
    0
  )
);
