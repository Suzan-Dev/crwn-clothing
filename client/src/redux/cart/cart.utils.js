export const addItemToCart = (cartItem, itemToAdd) => {
  const existingItem = cartItem.find((item) => item.id === itemToAdd.id);

  if (existingItem) {
    return cartItem.map((item) =>
      item.id === itemToAdd.id ? { ...item, quantity: item.quantity + 1 } : item
    );
  } else {
    return [...cartItem, { ...itemToAdd, quantity: 1 }];
  }
};

export const removeItemFromCart = (cartItem, itemToRemove) => {
  const existingItem = cartItem.find((item) => item.id === itemToRemove.id);

  if (existingItem.quantity === 1) {
    return cartItem.filter((item) => item.id !== itemToRemove.id);
  } else {
    return cartItem.map((item) =>
      item.id === itemToRemove.id
        ? { ...item, quantity: item.quantity - 1 }
        : item
    );
  }
};
