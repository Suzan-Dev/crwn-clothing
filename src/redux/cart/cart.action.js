import CartActionType from './cart.type';

export const ToggleHiddenCart = () => ({
  type: CartActionType.TOGGLE_HIDDEN_CART,
});

export const addItem = (item) => ({
  type: CartActionType.ADD_ITEM,
  payload: item,
});
