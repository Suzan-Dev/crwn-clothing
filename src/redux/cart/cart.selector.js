import { createSelector } from 'reselect';

const selectItem = (state) => state.cart;

export const selectCartItem = createSelector(
  [selectItem],
  (cart) => cart.cartItem
);

export const selectCartItemCount = createSelector(
  [selectCartItem],
  (cartItem) => cartItem.reduce((acc, cartItem) => acc + cartItem.quantity, 0)
);
