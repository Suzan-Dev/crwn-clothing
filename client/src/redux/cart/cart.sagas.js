import { all, put, call, takeLatest } from 'redux-saga/effects';
import UserActionType from '../user/user.type';
import { clearCart } from './cart.action';

// saga worker
export function* clearCartOnSuccess() {
  yield put(clearCart());
}

// saga watcher
export function* onSignOutSuccess() {
  yield takeLatest(UserActionType.SIGN_OUT_SUCCESS, clearCartOnSuccess);
}

export function* cartSagas() {
  yield all([call(onSignOutSuccess)]);
}
