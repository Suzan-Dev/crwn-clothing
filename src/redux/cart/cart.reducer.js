import CartActionType from './cart.type';
import { addItemToCart } from './cart.utils';

const INITIAL_STATE = {
  hidden: true,
  cartItem: [],
};

const CartReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CartActionType.TOGGLE_HIDDEN_CART:
      return {
        ...state,
        hidden: !state.hidden,
      };
    case CartActionType.ADD_ITEM:
      return {
        ...state,
        cartItem: addItemToCart(state.cartItem, action.payload),
      };
    default:
      return state;
  }
};

export default CartReducer;
