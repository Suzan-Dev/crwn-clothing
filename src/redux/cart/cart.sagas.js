import { all, put, call, takeLatest } from 'redux-saga/effects';

import UserActionType from '../user/user.type';

import { clearCart } from './cart.action';

// saga worker
function* clearCartOnSuccess() {
  yield put(clearCart());
}

// saga watcher
function* onSignOutSuccess() {
  yield takeLatest(UserActionType.SIGN_OUT_SUCCESS, clearCartOnSuccess);
}

export function* cartSagas() {
  yield all([call(onSignOutSuccess)]);
}
