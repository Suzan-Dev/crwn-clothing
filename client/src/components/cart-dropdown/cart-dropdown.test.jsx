import React from 'react';
import { shallow } from '../../setupTests';
import CartDropdown from './cart-dropdown.component';
import { ToggleHiddenCart } from '../../redux/cart/cart.action';

describe('Testing: CartDropdown component!', () => {
  const mockDispatch = jest.fn();
  const mockHistory = {
    push: jest.fn(),
  };
  const mockCartItem = [{ id: 1 }, { id: 2 }];

  const mockProps = {
    cartItem: mockCartItem,
    history: mockHistory,
    dispatch: mockDispatch,
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  test('should render CartDropdown component.', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should call history.push & dispatch ToggleHiddenCart action when button is clicked.', () => {
    wrapper.find('CustomButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(ToggleHiddenCart());
  });

  test('should render equal number of CartItem components as the cartItem props length.', () => {
    expect(wrapper.find('CartItem').length).toBe(mockCartItem.length);
  });

  test('should render empty message if cartItem is empty.', () => {
    const mockCartItem = [];
    const mockProps = {
      cartItem: mockCartItem,
      history: mockHistory,
      dispatch: mockDispatch,
    };
    const wrapper2 = shallow(<CartDropdown {...mockProps} />);
    expect(wrapper2.exists('.empty-message')).toBeTruthy();
  });
});
