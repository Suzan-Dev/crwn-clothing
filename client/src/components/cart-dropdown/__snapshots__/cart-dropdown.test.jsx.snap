// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing: CartDropdown component! should render CartDropdown component. 1`] = `
"<div className=\\"cart-dropdown\\">
  <div className=\\"cart-items\\">
    <CartItem item={{...}} />
    <CartItem item={{...}} />
  </div>
  <CustomButton onClick={[Function: onClick]}>
    <span className=\\"btn-text\\">
      Go To Checkout
    </span>
  </CustomButton>
</div>"
`;
